# Giriş

Bu bölümde bir matrisin izinin (trace of matices) ne olduğunu göreceğiz.

# 2.10  Trace Operatörü

<img src="https://drive.google.com/uc?id=1PKvrBkckqzCALZ4N7MwwLCqX3ZCZeyWG"  width="200" alt="Calculating the trace of a matrix" title="Calculating the trace of a matrix">

<em>The trace of matrix</em>


Trace, bir kare matrisin köşegenindeki tüm değerlerin toplamıdır.

$$
{A}=
\begin{bmatrix}
    2 & 9 & 8 \\\\
    4 & 7 & 1 \\\\
    8 & 2 & 5
\end{bmatrix}
$$

$$
\mathrm{Tr}({A}) = 2 + 7 + 5 = 14
$$

Numpy, onu hesaplamak için `trace()` işlevini sağlar:

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# Plot style
sns.set()
%pylab inline
pylab.rcParams['figure.figsize'] = (4, 4)

Populating the interactive namespace from numpy and matplotlib


In [ ]:
A = np.array([[2, 9, 8], [4, 7, 1], [8, 2, 5]])
A

array([[2, 9, 8],
       [4, 7, 1],
       [8, 2, 5]])

In [ ]:
A_tr = np.trace(A)
A_tr

14

GoodFellow ve ark. izinin bir matrisin Frobenius normunu belirtmek için kullanılabileceğini açıklar. Frobenius normu, matrisler için $L^2$ normunun eşdeğeridir. Şu şekilde tanımlanır:

$$
||{{A}}||_F=\sqrt{\sum_{i,j}A^2_{i,j}}
$$

Tüm elemanların karesini alın ve toplayın. Sonucun karekökünü alın. Bu norm ayrıca şu şekilde hesaplanabilir:

$$
||{{A}}||_F=\sqrt{Tr({{AA}^T})}
$$

Bunu kontrol edebiliriz. Normu hesaplamanın ilk yolu basit `np.linalg.norm()` komutuyla yapılabilir:

In [ ]:
np.linalg.norm(A)

17.549928774784245

${A}$'ın Frobenius normu 17.549928774784245'tir.

Trace ile sonuç aynıdır:

In [ ]:
np.sqrt(np.trace(A.dot(A.T)))

17.549928774784245

Bir matrisin tarnspozu köşegeni değiştirmediğinden, matrisin izi tanspozunun izine eşittir:


$$
Tr({A})=Tr({A}^T)
$$

## Çarpımın izi 

$$
Tr({ABC}) = Tr({CAB}) = Tr({BCA})
$$


### Örnek 1.

Bu özelliğin bir örneğini görelim.

$$
{A}=
\begin{bmatrix}
    4 & 12 \\\\
    7 & 6
\end{bmatrix}
$$

$$
{B}=
\begin{bmatrix}
    1 & -3 \\\\
    4 & 3
\end{bmatrix}
$$

$$
{C}=
\begin{bmatrix}
    6 & 6 \\\\
    2 & 5
\end{bmatrix}
$$

In [ ]:
A = np.array([[4, 12], [7, 6]])
B = np.array([[1, -3], [4, 3]])
C = np.array([[6, 6], [2, 5]])

np.trace(A.dot(B).dot(C))

531

In [ ]:
np.trace(C.dot(A).dot(B))

531

In [ ]:
np.trace(B.dot(C).dot(A))

531

$$
{ABC}=
\begin{bmatrix}
    360 & 432 \\\\
    180 & 171
\end{bmatrix}
$$

$$
{CAB}=
\begin{bmatrix}
    498 & 126 \\\\
    259 & 33
\end{bmatrix}
$$

$$
{BCA}=
\begin{bmatrix}
    -63 & -54 \\\\
    393 & 594
\end{bmatrix}
$$

$$
Tr({ABC}) = Tr({CAB}) = Tr({BCA}) =  531
$$

# Referanslar

[Trace (linear algebra) - Wikipedia](https://en.wikipedia.org/wiki/Trace_(linear_algebra))

[Numpy Trace operator](https://docs.scipy.org/doc/numpy/reference/generated/numpy.trace.html)